File: python/text/ciphers.py
"""Vigenere / Vernam / Caesar Ciphers — classical text encryption."""


def caesar_encrypt(text: str, shift: int) -> str:
    """Encrypt text using Caesar cipher with given shift."""
    result = []
    shift = shift % 26
    for ch in text:
        if ch.isalpha():
            base = ord('A') if ch.isupper() else ord('a')
            result.append(chr((ord(ch) - base + shift) % 26 + base))
        else:
            result.append(ch)
    return ''.join(result)


def caesar_decrypt(text: str, shift: int) -> str:
    """Decrypt Caesar cipher by shifting in the opposite direction."""
    return caesar_encrypt(text, -shift)


def vigenere_encrypt(text: str, key: str) -> str:
    """Encrypt text using Vigenere cipher with given keyword."""
    if not key or not key.isalpha():
        raise ValueError("Key must be a non-empty alphabetic string")
    key_upper = key.upper()
    result = []
    ki = 0
    for ch in text:
        if ch.isalpha():
            shift = ord(key_upper[ki % len(key_upper)]) - ord('A')
            base = ord('A') if ch.isupper() else ord('a')
            result.append(chr((ord(ch) - base + shift) % 26 + base))
            ki += 1
        else:
            result.append(ch)
    return ''.join(result)


def vigenere_decrypt(text: str, key: str) -> str:
    """Decrypt Vigenere cipher by inverting each key letter's shift."""
    if not key or not key.isalpha():
        raise ValueError("Key must be a non-empty alphabetic string")
    inverted_key = ''.join(
        chr((26 - (ord(c) - ord('A'))) % 26 + ord('A')) for c in key.upper()
    )
    return vigenere_encrypt(text, inverted_key)


def vernam_encrypt(data: bytes, key: bytes) -> bytes:
    """Encrypt bytes using Vernam (one-time pad) XOR cipher."""
    if len(key) < len(data):
        raise ValueError("Key must be at least as long as the data")
    return bytes(d ^ k for d, k in zip(data, key))


def vernam_decrypt(data: bytes, key: bytes) -> bytes:
    """Decrypt Vernam cipher (XOR is its own inverse)."""
    return vernam_encrypt(data, key)


if __name__ == "__main__":
    # Caesar demo
    plain = "Hello, World!"
    shift = 13
    enc = caesar_encrypt(plain, shift)
    dec = caesar_decrypt(enc, shift)
    print(f"Caesar (shift={shift}):")
    print(f"  Plain:     {plain}")
    print(f"  Encrypted: {enc}")
    print(f"  Decrypted: {dec}")
    assert dec == plain, "Caesar roundtrip failed"
    print()

    # Vigenere demo
    key = "SECRET"
    enc = vigenere_encrypt(plain, key)
    dec = vigenere_decrypt(enc, key)
    print(f"Vigenere (key={key!r}):")
    print(f"  Plain:     {plain}")
    print(f"  Encrypted: {enc}")
    print(f"  Decrypted: {dec}")
    assert dec == plain, "Vigenere roundtrip failed"
    print()

    # Vernam demo
    message = b"Hello, World!"
    otp_key = b"\x4a\x1f\x7c\x33\x0b\x5e\x22\x6d\x11\x44\x59\x08\x3a"
    enc = vernam_encrypt(message, otp_key)
    dec = vernam_decrypt(enc, otp_key)
    print(f"Vernam (one-time pad):")
    print(f"  Plain:     {message}")
    print(f"  Encrypted: {enc.hex()}")
    print(f"  Decrypted: {dec}")
    assert dec == message, "Vernam roundtrip failed"
    print()

    print("All roundtrip tests passed.")

File: rust/text/ciphers.rs
//! Vigenere / Vernam / Caesar Ciphers — classical text encryption.

/// Encrypt text using Caesar cipher with given shift.
pub fn caesar_encrypt(text: &str, shift: i32) -> String {
    let s = ((shift % 26) + 26) as u32 % 26;
    text.chars()
        .map(|ch| {
            if ch.is_ascii_alphabetic() {
                let base = if ch.is_ascii_uppercase() { b'A' } else { b'a' } as u32;
                char::from_u32((ch as u32 - base + s) % 26 + base).unwrap()
            } else {
                ch
            }
        })
        .collect()
}

/// Decrypt Caesar cipher by shifting in the opposite direction.
pub fn caesar_decrypt(text: &str, shift: i32) -> String {
    caesar_encrypt(text, -shift)
}

/// Encrypt text using Vigenere cipher with given keyword.
pub fn vigenere_encrypt(text: &str, key: &str) -> String {
    assert!(
        !key.is_empty() && key.chars().all(|c| c.is_ascii_alphabetic()),
        "Key must be a non-empty alphabetic string"
    );
    let key_bytes: Vec<u8> = key.to_ascii_uppercase().bytes().collect();
    let mut ki = 0usize;
    text.chars()
        .map(|ch| {
            if ch.is_ascii_alphabetic() {
                let shift = (key_bytes[ki % key_bytes.len()] - b'A') as u32;
                let base = if ch.is_ascii_uppercase() { b'A' } else { b'a' } as u32;
                ki += 1;
                char::from_u32((ch as u32 - base + shift) % 26 + base).unwrap()
            } else {
                ch
            }
        })
        .collect()
}

/// Decrypt Vigenere cipher by inverting each key letter's shift.
pub fn vigenere_decrypt(text: &str, key: &str) -> String {
    assert!(
        !key.is_empty() && key.chars().all(|c| c.is_ascii_alphabetic()),
        "Key must be a non-empty alphabetic string"
    );
    let inverted: String = key
        .to_ascii_uppercase()
        .bytes()
        .map(|b| ((26 - (b - b'A') as u32) % 26) as u8 + b'A')
        .map(|b| b as char)
        .collect();
    vigenere_encrypt(text, &inverted)
}

/// Encrypt bytes using Vernam (one-time pad) XOR cipher.
/// Key must be at least as long as data.
pub fn vernam_encrypt(data: &[u8], key: &[u8]) -> Vec<u8> {
    assert!(
        key.len() >= data.len(),
        "Key must be at least as long as the data"
    );
    data.iter().zip(key.iter()).map(|(d, k)| d ^ k).collect()
}

/// Decrypt Vernam cipher (XOR is its own inverse).
pub fn vernam_decrypt(data: &[u8], key: &[u8]) -> Vec<u8> {
    vernam_encrypt(data, key)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_caesar_roundtrip() {
        let plain = "Hello, World!";
        for shift in [0, 1, 13, 25, -3] {
            let enc = caesar_encrypt(plain, shift);
            let dec = caesar_decrypt(&enc, shift);
            assert_eq!(dec, plain, "Caesar roundtrip failed for shift={shift}");
        }
    }

    #[test]
    fn test_caesar_preserves_non_alpha() {
        let text = "abc 123 !@#";
        let enc = caesar_encrypt(text, 5);
        assert_eq!(&enc[3..], " 123 !@#");
    }

    #[test]
    fn test_caesar_preserves_case() {
        let enc = caesar_encrypt("AbCd", 1);
        assert_eq!(enc, "BcDe");
    }

    #[test]
    fn test_vigenere_roundtrip() {
        let plain = "Hello, World!";
        let key = "SECRET";
        let enc = vigenere_encrypt(plain, key);
        let dec = vigenere_decrypt(&enc, key);
        assert_eq!(dec, plain);
    }

    #[test]
    fn test_vigenere_known_value() {
        // A with shift S(18) => S
        let enc = vigenere_encrypt("A", "S");
        assert_eq!(enc, "S");
    }

    #[test]
    fn test_vernam_roundtrip() {
        let data = b"Hello, World!";
        let key = b"\x4a\x1f\x7c\x33\x0b\x5e\x22\x6d\x11\x44\x59\x08\x3a";
        let enc = vernam_encrypt(data, key);
        let dec = vernam_decrypt(&enc, key);
        assert_eq!(dec, data);
    }

    #[test]
    #[should_panic(expected = "Key must be at least as long")]
    fn test_vernam_short_key() {
        vernam_encrypt(b"Hello", b"Hi");
    }
}
